### PR TITLE
[Kotlin Cleanup] DeckAdapter

### PR DESCRIPTION
**Pull request by dorrin-sot**

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt [hunk 1]
```diff
@@ -285,18 +286,17 @@ class DeckAdapter(private val layoutInflater: LayoutInflater, context: Context)
      * An invalid deck ID will return position 0.
      */
     fun findDeckPosition(did: Long): Int {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:292]: Can we add unit tests if we're seriously refactoring. This adds risk for little reward otherwise.

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt [hunk 2]
```diff
@@ -253,35 +253,34 @@ class DeckAdapter(private val layoutInflater: LayoutInflater, context: Context)
     }
 
     private fun processNodes(nodes: List<TreeNode<AbstractDeckTreeNode>>) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:255]: unit test please

**Review by david-allison — CHANGES_REQUESTED** [review]
In general, looks much ore Kotlin-y, thanks!

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt (hunk 1 above)
Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:291]: I removed the bit at the end since it was causing stackoverflow in tests; so only the `for` loop is getting cleaned up. Does it still need unit test?

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt (hunk 2 above)
Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:254]: Is it ok if I just add a `@NeedsTest` because I have no idea what this method does or how to test it :(

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt (hunk 1 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:291]: I want to encourage refactoring tested code, or adding tests while refactoring. But: use your judgment.

Everyone has different ideas about style. The most important thing to me is that we document the functions, and add tests, and make sure that we're not introducing regressions (such as the stack overflow). Those are less likely to change than the 'style' of the code, and to an extent, they'll help enable future refactorings.

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt (hunk 2 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:254]: Hardline stance: if you don't know what the method did beforehand, it shouldn't be refactored, as you don't know whether it's doing the same thing now. 

More time should be spent on understanding it, and documentation/documentation via tests/readability refactorings should be added first.

This is increasing the risk of bugs for a code style change. It's not a good tradeoff to take.

Note: I haven't touched this method at all, and also would need some time to understand it.

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:254]: Still pending

Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:254]: Oh In that case, can I drop the commit related to this method?

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:254]: Either or: understand and document, or drop. 

Dropping would be better for velocity if you have other plans

Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:254]: As you mentioned it's too much for little payoff (it's not that much cleaner. I'd go as far as to say  in some parts it's borderline harder to read even😂)

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt (hunk 1 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:292]: Don't use exceptions for non-exceptional circumstances.

Extract `mCol!!.decks.parents(did).last().optLong("id", 0)` to `DeckService` and document it, rather than trying to make it more terse.

We should be testing functionality that we're refactoring, otherwise we're introducing risk


I'd imagine this would be:

```kotlin
    /**
     * Return the position of the deck in the deck list. If the deck is a child of a collapsed deck
     * (i.e., not visible in the deck list), then the position of the parent deck is returned instead.
     *
     * An invalid deck ID will return position 0.
     */
    fun findDeckPosition(did: Long): Int {
        return mCurrentDeckList.indices
            .find { mCurrentDeckList[it].value.did == did }
             // If the deck is not in our list, we search again using the immediate parent
            ?: getParentDid(did)?.let { parentDid -> findDeckPosition(parentDid) } ?: 0
    }
```

**Comment by dorrin-sot** [comment]
> > [Merge branch 'main' into Kotlin_Cleanup_DeckAdapter](https://github.com/ankidroid/Anki-Android/pull/11634/commits/accf615905f443a28f16fd55ece841bfceb8114f)
> 
> 😱

What happenned 😳

**Comment by david-allison** [comment]
Hi, please remove the merge commit and rebase onto `main` instead.

**Comment by dorrin-sot** [comment]
Oh sorry done😅

File: AnkiDroid/src/main/java/com/ichi2/libanki/utils/IterableUtils.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ ****************************************************************************************/
+
+package com.ichi2.libanki.utils
+
+object IterableUtils {
+    /**
+     * @param element the element to search iterable for
+     * @return the index of [element]. returns null if none found
+     */
+    fun <T> Iterable<T>.nullableIndexOf(element: T): Int? {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/libanki/utils/IterableUtils.kt:24]: This can be a bare function, outside the object.

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt [hunk 4]
```diff
@@ -286,34 +292,21 @@ class DeckAdapter(private val layoutInflater: LayoutInflater, context: Context)
...
-        // If the deck is not in our list, we search again using the immediate parent
-        val parents = mCol.decks.parents(did)
-        return if (parents.isEmpty()) {
-            0
-        } else {
-            findDeckPosition(parents[parents.size - 1].optLong("id", 0))
-        }
+        return deckList
+            .map { it.value.did }
+            .nullableIndexOf(did)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:297]: The standard Kotlin suffix is `OrNull`. So `indexOfOrNull`. Kotlin already defines this on `String`.

I don't see why this wasn't shortened to `.indexOfOrNull { it.value.did == did }

**Review by david-allison — CHANGES_REQUESTED** [review]
A few questions on style

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:293]: Oh, I see it was actually changed in later commit. Would you mind doing an interactive rebase and do the change as expected in the first commit changing those lines, so it gets easier to review

File: AnkiDroid/src/main/java/com/ichi2/libanki/utils/IterableUtils.kt (hunk 3 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/utils/IterableUtils.kt]: Please don't resolve, it seems to be still inside object.

Also, it'd be great two get tests. Just one with null and one with a number would be sufficient

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt [hunk 5]
```diff
@@ -296,17 +296,10 @@ class DeckAdapter(private val layoutInflater: LayoutInflater, context: Context)
     }
 
     val eta: Int?
-        get() = if (mNumbersComputed) {
-            mCol.sched.eta(Counts(mNew, mLrn, mRev))
-        } else {
-            null
-        }
+        get() = mCol.sched.eta(Counts(mNew, mLrn, mRev)).takeIf { mNumbersComputed }
```
[the remark this answers is not in the document]
Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt]: Hi I fixed this and https://github.com/ankidroid/Anki-Android/pull/11634#discussion_r901120215 in later commits

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt [hunk 6]
```diff
@@ -336,8 +336,8 @@ class DeckAdapter(private val layoutInflater: LayoutInflater, context: Context)
         }
 
         private fun containsFilterString(filterPattern: String, root: AbstractDeckTreeNode): Boolean {
-            val deckName = root.fullDeckName
-            return deckName.lowercase(Locale.getDefault()).contains(filterPattern) || deckName.lowercase(Locale.ROOT).contains(filterPattern)
+            return listOf(Locale.getDefault(), Locale.ROOT)
```
[the remark this answers is not in the document]
Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:345]: Finally kotlin added these🤩 I converted all the "convertable into expression body" functions

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt]: I just thought of this but I'm not sure how I feel about it
```
val eta: Int?
    get() = mNumbersComputed
        .takeIf { it }
        ?.let { mCol.sched.eta(Counts(mNew, mLrn, mRev)) }
```

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt [hunk 7]
```diff
@@ -150,24 +150,26 @@ class DeckAdapter(private val layoutInflater: LayoutInflater, context: Context)
...
-            deckLayout.setPadding(smallPadding, 0, rightPadding, 0)
-            holder.deckExpander.visibility = View.VISIBLE
             // Create the correct expander for this deck
             setDeckExpander(holder.deckExpander, holder.indentView, treeNode)
-        } else {
-            holder.deckExpander.visibility = View.GONE
-            val normalPadding = deckLayout.resources.getDimension(R.dimen.deck_picker_left_padding).toInt()
-            deckLayout.setPadding(normalPadding, 0, rightPadding, 0)
         }
+        val leftPadding = deckLayout.resources.getDimension(
```
[the remark this answers is not in the document]
Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:155]: I tried running with persian too. the values of `leftPadding` and `rightPadding` are the same. but also the padding amounts look the same. so maybe RTL is being handled internally (inside android's `RelativeLayout`)?🤔

<div style="display: flex; flex-direction: row">
image
image
</div>

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt [hunk 8]
```diff
@@ -150,24 +150,26 @@ class DeckAdapter(private val layoutInflater: LayoutInflater, context: Context)
         // Set the expander icon and padding according to whether or not there are any subdecks
         val deckLayout = holder.deckLayout
         val rightPadding = deckLayout.resources.getDimension(R.dimen.deck_picker_right_padding).toInt()
+
         if (mHasSubdecks) {
-            val smallPadding = deckLayout.resources.getDimension(R.dimen.deck_picker_left_padding_small).toInt()
-            deckLayout.setPadding(smallPadding, 0, rightPadding, 0)
-            holder.deckExpander.visibility = View.VISIBLE
```
[the remark this answers is not in the document]
Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:157]: Oh sorry I missed it

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DeckService.kt [hunk 9]
```diff
@@ -58,4 +58,11 @@ object DeckService {
             "select count() from cards where did in $ids or odid in $ids"
         )
     }
+
+    @JvmStatic
+    fun getParentDid(col: Collection, did: Long): Long? {
+        return col.decks.parents(did)
+            .takeUnless { pList -> pList.isEmpty() }
+            ?.run { last().optLong("id", 0) }
+    }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DeckService.kt:67]: Can this use `lastOrNull`: 

```kotlin
return col.decks.parents(did).lastOrNull()?.optLong("id", 0)
```

**Review by david-allison — COMMENTED** [review]
LGTM. One potential further cleanup

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DeckService.kt (hunk 9 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DeckService.kt]: Why was this changed to use `.run`?

**Comment by dorrin-sot** [comment]
also how can I fix a translation string I just noticed was mistranslated in [this comment](https://github.com/ankidroid/Anki-Android/pull/11634#discussion_r901914986)? can I just change the string xml file?

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt [hunk 10]
```diff
@@ -331,17 +321,9 @@ class DeckAdapter(private val layoutInflater: LayoutInflater, context: Context)
...
         }
     val due: Int?
-        get() = if (mNumbersComputed) {
-            mNew + mLrn + mRev
-        } else {
-            null
-        }
-    private val deckList: List<TreeNode<AbstractDeckTreeNode>>
-        get() = mCurrentDeckList
+        get() = (mNew + mLrn + mRev).takeIf { mNumbersComputed }
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:324]: I'm not fan of `takeIf` here since it means the work is done before the condition is checked. 
It's a sum, so not such a big deal. But I needed to note it at least

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt [hunk 11]
```diff
@@ -380,14 +357,22 @@ class DeckAdapter(private val layoutInflater: LayoutInflater, context: Context)
...
     companion object {
         /* Make the selected deck roughly half transparent if there is a background */
         const val SELECTED_DECK_ALPHA_AGAINST_BACKGROUND = 0.45
+
+        /**
+         * @param predicate the predicate to find index of
+         * @return the index of [predicate]. returns null if none found
+         */
+        @VisibleForTesting
+        fun <T> Iterable<T>.indexOfOrNull(predicate: (T) -> Boolean): Int? =
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:374]: I'd move it to com.ichi2.utils.CollectionUtils if it's okay

File: AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (dorrin-sot) [reply AnkiDroid/src/main/java/com/ichi2/anki/widgets/DeckAdapter.kt:158]: Is it ok if I set them to "start padding" and "end padding" since in xml layouts the android's padding attributes are named like this?
